frb_codegen/src/utils.rs

```diff
@@ -83,3 +89,196 @@ impl Display for BlockIndex {
         write!(f, "{}", self.0)
     }
 }
+
+#[derive(Debug, PartialEq)]
+pub(crate) enum DartToolchain {
+    Dart,
+    Flutter,
+}
+
+impl DartToolchain {
+    pub(crate) fn as_run_command(&self) -> &'static str {
+        match self {
+            DartToolchain::Dart => "dart",
+            DartToolchain::Flutter => "flutter pub",
+        }
+    }
+}
+
+#[derive(Debug, Deserialize)]
+struct PubspecLock {
+    pub packages: HashMap<String, serde_yaml::Value>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(untagged)]
+pub enum PackageVersion {
+    Inline(String),
+    Multiline { version: Option<String> },
+}
+
+#[derive(Debug)]
+pub enum PackageVersionKind {
+    Exact(Version),
+    Range(VersionReq),
+}
+
+impl TryFrom<&PackageVersion> for PackageVersionKind {
+    type Error = anyhow::Error;
+    fn try_from(version: &PackageVersion) -> Result<Self, Self::Error> {
+        match version {
+            PackageVersion::Inline(ref version) => {
+                let version = PackageVersionKind::from_str(version)?;
+                Ok(version)
+            }
+            PackageVersion::Multiline { ref version } => {
+                if let Some(version) = version {
+                    let version = PackageVersionKind::from_str(version)?;
+                    return Ok(version);
+                }
+                Err(anyhow::anyhow!("no version found"))
+            }
+        }
+    }
```

Comment:
nit: Maybe we can extract a new function: `impl PackageVersion { fn version(&self) -> String {}}`. Then, in this function, we no longer need to `match version`

Reply:
done in d56544d94a540aa5b3143812c8c8ee0f60d96369